# Imports

In [1]:
## Suppress warnings
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
import numpy as np
import seaborn as sns
import pickle

## Notebook display settings
from IPython.display import display
pd.options.display.max_columns = None
pd.options.display.max_rows = None
pd.options.display.max_seq_items = None

# Own scripts
from functions import get_start_end_points, generate_histogram2d, gen_spirals, analyse_spiral
## Plot style
%matplotlib inline

# Daten

In [2]:
with open('data.pickle', 'rb') as handle:
    data = pickle.load(handle)
data.head()

id  run               timestamp  segment          start_time  \
0  163.0  0.0 2017-05-30 09:21:13.150     11.0 2017-05-30 09:21:05   
1  164.0  0.0 2017-05-30 09:21:13.200     11.0 2017-05-30 09:21:05   
2  165.0  0.0 2017-05-30 09:21:13.250     11.0 2017-05-30 09:21:05   
3  166.0  0.0 2017-05-30 09:21:13.300     11.0 2017-05-30 09:21:05   
4  167.0  0.0 2017-05-30 09:21:13.350     11.0 2017-05-30 09:21:05   

                 end_time   force_x   force_y   force_z       pos_x  \
0 2017-05-30 09:22:41.607 -0.543045 -2.924596 -7.257000 -257.695704   
1 2017-05-30 09:22:41.607 -0.465969 -3.278856 -7.522000 -257.617629   
2 2017-05-30 09:22:41.607 -0.317475 -3.435834 -7.705999 -256.626803   
3 2017-05-30 09:22:41.607 -0.298383 -3.457754 -7.580999 -254.594017   
4 2017-05-30 09:22:41.607 -0.289898 -3.447854 -7.539000 -251.411919   

        pos_y       pos_z       rot_x       rot_y       rot_z result_robot  
0 -265.980105  240.007357   -0.008832    0.034723  179.958414      Success  
1 -265.999590  240.004745   -0.000171    0.019928  179.968161      Success  
2 -266.602650  239.791224 -179.993978  179.989304    0.164766      Success  
3 -267.926090  239.323013 -179.992478  179.999036    0.577883      Success  
4 -269.894497  238.550625 -179.996649  179.988091    1.210269      Success

# Code

## Listen

Start und Endpunkte der Such-Segmente finden

In [3]:
start_points, end_points, rel_end_points = get_start_end_points(data)

In [4]:
hist, xedge, yedge = generate_histogram2d(rel_end_points)

# Spiralen

In [5]:
# 1 cluster
spirals_k1 = gen_spirals(hist, xedge, yedge, 0.028, -0.042)

# 2 cluster
spirals_k2_c1 = gen_spirals(hist, xedge, yedge, 0.020, -0.066)
spirals_k2_c2 = gen_spirals(hist, xedge, yedge, 0.092, 0.139)


## Definitionen

Zur besseren Verständlichkeit soll folgendes gelten:

- Norden (N) und Osten (E) sind die jeweiligen positiven Richtungen der x oder y Achse
- Süden (S) und Westen (W) sind die jeweiligen negativen Richtungen der x oder y Achse
- Die x Achse wird horizontal dargestellt, die Werte werden von niedrig/negativ (links) nach hoch/positiv (rechts) aufgetragen
- Die y Achse wird vertikal dargestellt, die Werte werden von niedrig/negativ (unten) nach hoch/positiv (oben) aufgetragen

## Analyse

### Ein Cluster

In [6]:
_ = analyse_spiral(spirals_k1, hist)

direction.S turn.turn_left: 60864.0

direction.S turn.turn_right: 59888.0

direction.N turn.turn_left: 60536.0

direction.N turn.turn_right: 61432.0

direction.W turn.turn_left: 60168.0

direction.W turn.turn_right: 61112.0

direction.E turn.turn_left: 61248.0

direction.E turn.turn_right: 60928.0

Best result: 
direction.S turn.turn_right
59888.0
[[24  9 10 11 12]
 [23  8  1  2 13]
 [22  7  0  3 14]
 [21  6  5  4 15]
 [20 19 18 17 16]]


Mit der relativen Verschiebung des Startpunktes um ~(0.028, -0.042) erhalten wir eine Verbesserung des Scores zwischen den besten Spiralsuchen auf 59888 von 67376 und somit eine prozentuale Verbesserung von ~11%

### Zwei Cluster

In [7]:
_ = analyse_spiral(spirals_k2_c1, hist)

direction.S turn.turn_left: 65450.0

direction.S turn.turn_right: 62728.0

direction.N turn.turn_left: 63078.0

direction.N turn.turn_right: 65852.0

direction.W turn.turn_left: 62564.0

direction.W turn.turn_right: 63102.0

direction.E turn.turn_left: 65512.0

direction.E turn.turn_right: 65546.0

Best result: 
direction.W turn.turn_left
62564.0
[[16 15 14 13 12]
 [17  4  3  2 11]
 [18  5  0  1 10]
 [19  6  7  8  9]
 [20 21 22 23 24]]


In [8]:
_ = analyse_spiral(spirals_k2_c2, hist)

direction.S turn.turn_left: 239642.0

direction.S turn.turn_right: 266074.0

direction.N turn.turn_left: 261506.0

direction.N turn.turn_right: 243154.0

direction.W turn.turn_left: 273326.0

direction.W turn.turn_right: 277294.0

direction.E turn.turn_left: 250310.0

direction.E turn.turn_right: 254966.0

Best result: 
direction.S turn.turn_left
239642.0
[[20 19 18 17 16]
 [21  6  5  4 15]
 [22  7  0  3 14]
 [23  8  1  2 13]
 [24  9 10 11 12]]


Das Ergebnis des zweiten Zentrums (0.092, 0.139) ergibt einen zu Erwartenden deutlich schlechteren Wert.  
  
Mit der relativen Verschiebung des Startpunktes um ~(0.020, -0.066) erhalten wir eine Verbesserung des Scores zwischen den besten Spiralsuchen auf 62564 von 67376 und somit eine prozentuale Verbesserung von ~7%. Dieser Wert ist identisch zum besten Wert der euklidischen Distanz, nutzt aber eine andere Spiralsuche.